ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -368,91 +364,69 @@ private void assignReplicasToThreadPool() throws IOException {
-                      && remoteReplicaInfo.getReplicaId().getReplicaPath().equals(replicaPath)) {
-                    logger.info("Read token for partition {} remote host {} port {} token {}", partitionId, hostname,
-                        port, token);
-                    if (!partitionInfo.getStore().isEmpty()) {
-                      remoteReplicaInfo.initializeTokens(token);
-                      remoteReplicaInfo.setTotalBytesReadFromLocalStore(totalBytesReadFromLocalStore);
-                    } else {
-                      // if the local replica is empty, it could have been newly created. In this case, the offset in
-                      // every peer replica which the local replica lags from should be set to 0, so that the local
-                      // replica starts fetching from the beginning of the peer. The totalBytes the peer read from the
-                      // local replica should also be set to 0. During initialization these values are already set to 0,
-                      // so we let them be.
-                      tokenWasReset = true;
-                      logTokenReset(partitionId, hostname, port, token);
-                    }
-                    updatedToken = true;
-                    break;
-                  }
-                }
-                if (!updatedToken) {
-                  logger.warn("Persisted remote replica host {} and port {} not present in new cluster ", hostname,
-                      port);
-                }
+    long readStartTimeMs = SystemTime.getInstance().milliseconds();
+    try {
+      List<ReplicaTokenInfo> tokenInfoList = persistor.retrieveTokens(mountPath);
+
+      for (ReplicaTokenInfo tokenInfo : tokenInfoList) {
+        String hostname = tokenInfo.getHostname();
+        int port = tokenInfo.getPort();
+        PartitionId partitionId = tokenInfo.getPartitionId();
+        FindToken token = tokenInfo.getReplicaToken();
+        // update token
+        PartitionInfo partitionInfo = partitionsToReplicate.get(tokenInfo.getPartitionId());
+        if (partitionInfo != null) {
+          boolean updatedToken = false;
+          for (RemoteReplicaInfo remoteReplicaInfo : partitionInfo.getRemoteReplicaInfos()) {
+            DataNodeId dataNodeId = remoteReplicaInfo.getReplicaId().getDataNodeId();
+            if (dataNodeId.getHostname().equalsIgnoreCase(hostname) && dataNodeId.getPort() == port && remoteReplicaInfo
+                .getReplicaId()
+                .getReplicaPath()
+                .equals(tokenInfo.getReplicaPath())) {
+              logger.info("Read token for partition {} remote host {} port {} token {}", partitionId, hostname, port,
+                  token);
+              if (!partitionInfo.getStore().isEmpty()) {
+                remoteReplicaInfo.initializeTokens(token);
+                remoteReplicaInfo.setTotalBytesReadFromLocalStore(tokenInfo.getTotalBytesReadFromLocalStore());
```

Comment:
Can we combine these two call to one? 
tokenInfo can provide all info. 

Also, since you introduced `ReplicaTokenInfo`. I think it's good to hide `FindToken` from `remoteReplicaInfo` constructor and interfaces. 
For constructor, you can just pass in a findTokenFactory.

Reply:
Is it possible to move both `ReplicaTokenInfo` and  `ReplicaTokeninfo` to `RemoteReplicaInfo` as inner static classes?

Reply:
Move ReplicaTokenInfo and ???

Reply:
Per second comment, I'm not sure what your suggesting to do with the RemoteReplicaInfo constructor, and what that has to do with the ReplicaTokenInfo.  Passing token vs factory seems independent.

Reply:
Sorry. I originally mean move both `ReplicaTokenInfo` and `ReplicaTokenSerde` to `RemoteReplicaInfo` as inner static classes.
------
How about just move `ReplicaTokenInfo` to `RemoteReplicaInfo`.

Reply:
Made ReplicaTokenInfo a static inner class of RemoteReplicaInfo.